src/github.com/matrix-org/dendrite/cmd/dendrite-monolith-server/main.go

```diff
@@ -0,0 +1,278 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"net/http"
+	"os"
+
+	"github.com/gorilla/mux"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/devices"
+	"github.com/matrix-org/dendrite/clientapi/producers"
+	"github.com/matrix-org/dendrite/common"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/dendrite/common/keydb"
+	"github.com/matrix-org/gomatrixserverlib"
+
+	mediaapi_routing "github.com/matrix-org/dendrite/mediaapi/routing"
+	mediaapi_storage "github.com/matrix-org/dendrite/mediaapi/storage"
+
+	roomserver_alias "github.com/matrix-org/dendrite/roomserver/alias"
+	roomserver_input "github.com/matrix-org/dendrite/roomserver/input"
+	roomserver_query "github.com/matrix-org/dendrite/roomserver/query"
+	roomserver_storage "github.com/matrix-org/dendrite/roomserver/storage"
+
+	clientapi_consumers "github.com/matrix-org/dendrite/clientapi/consumers"
+	clientapi_routing "github.com/matrix-org/dendrite/clientapi/routing"
+
+	syncapi_consumers "github.com/matrix-org/dendrite/syncapi/consumers"
+	syncapi_routing "github.com/matrix-org/dendrite/syncapi/routing"
+	syncapi_storage "github.com/matrix-org/dendrite/syncapi/storage"
+	syncapi_sync "github.com/matrix-org/dendrite/syncapi/sync"
+	syncapi_types "github.com/matrix-org/dendrite/syncapi/types"
+
+	federationapi_routing "github.com/matrix-org/dendrite/federationapi/routing"
+
+	federationsender_consumers "github.com/matrix-org/dendrite/federationsender/consumers"
+	"github.com/matrix-org/dendrite/federationsender/queue"
+	federationsender_storage "github.com/matrix-org/dendrite/federationsender/storage"
+
+	log "github.com/Sirupsen/logrus"
+	sarama "gopkg.in/Shopify/sarama.v1"
+)
+
+var (
+	logDir        = os.Getenv("LOG_DIR")
+	configPath    = flag.String("config", "dendrite.yaml", "The path to the config file. For more information, see the config file in this repository.")
+	httpBindAddr  = flag.String("http-bind-address", ":8008", "The HTTP listening port for the server")
+	httpsBindAddr = flag.String("https-bind-address", ":8448", "The HTTPS listening port for the server")
+)
+
+func main() {
+	common.SetupLogging(logDir)
+
+	flag.Parse()
+
+	if *configPath == "" {
+		log.Fatal("--config must be supplied")
+	}
+	cfg, err := config.Load(*configPath)
+	if err != nil {
+		log.Fatalf("Invalid config file: %s", err)
+	}
+
+	m := monolith{cfg: cfg, api: mux.NewRouter()}
```

Comment:
Even though this is a private type, it's still important to tell the reader which fields need to be supplied to `monolith` and which it will set when you call `setupXXX`, without having to read 200 lines of monolith functions. I'd be happier with one of:
 - Add a small `newMonolith(cfg, api)` unexported function.
 - Re-jig the order and grouping of fields in the type definition and add a one-liner on the required fields.